bindings/rust/s2n-tls-tokio/src/lib.rs

```diff
@@ -89,10 +105,26 @@ where
     type Output = Result<(), Error>;
 
     fn poll(mut self: Pin<&mut Self>, ctx: &mut Context<'_>) -> Poll<Self::Output> {
-        self.tls.with_io(ctx, |context| {
-            let conn = context.get_mut().as_mut();
-            conn.negotiate().map(|r| r.map(|_| ()))
-        })
+        let result = match self.error.take() {
```

Comment:
This future might get called before `poll_shutdown` completes. I think we need to call that here if `self.error.is_some()`

Reply:
I'm not sure I understand this comment-- isn't that what I'm doing? If there's an error, we don't call negotiate again, we just jump down to checking whether or not we need to call shutdown.

Would it be clearer if I didn't `take` the error? I'm realizing that's probably unnecessary, since we don't need to erase the error after shutdown completes, because that's the end of the future and so won't be polled again.

Reply:
No I'm calling out the following scenario:

```rust
// this encounters an error, calls poll_shutdown, stores error in self.error, returns Pending
handshake.poll();
// I immediately call poll again, now it returns `Err(err)` before `poll_shutdown` has returned `Ready`
handshake.poll();
```

`Future::poll` implementations should be able to be called any time and do the right thing. Consider the following join combinator:

```rust
struct Join<A, B> {
    a: A,
    b: B,
}

impl<A: Future, B: Future> Future for Join<A, B> {
    type Output = (A::Output, B::Output);

    fn poll(self: Pin<&mut Self>, cx: &mut Context) -> Poll<Self::Output> {
        // make progress on both sub-futures (these futures should be fused, but we'll ignore that here)
        let a = Pin::new(&mut self.a).poll(cx);
        let b = Pin::new(&mut self.b).poll(cx);
        // return if either result is pending
        let a = ready!(a);
        let b = ready!(b);
        // return the joined result
        Poll::Ready((a, b))
    }
}
```

If _either_ A or B call `waker.wake()`, then this future will call `poll` for _both_ A and B, even if the other one isn't quite ready to be woken.